components/client/src/test/unit/java/com/hotels/styx/client/connectionpool/SimpleConnectionPoolTest.java

```diff
@@ -921,4 +922,98 @@ public void ensureInEstablishmentCountIsDecrementedInError() throws InterruptedE
+        StepVerifier.create(pool.borrowConnection())
+                .expectNext(connection1)
+                .verifyComplete();
+        assertEquals(pool.stats().connectionAttempts(), 1);
+        assertEquals(pool.stats().busyConnectionCount(), 1);
+
+        pool.close();
+        assertEquals(0, pool.stats(). closedConnections());
+
+        doAnswer( arg -> {
+            when(connection1.isConnected()).thenReturn(false);
+            return true;
+        }).when(connection1).close();
+        pool.returnConnection(connection1);
+        assertEquals(pool.stats().connectionAttempts(), 1);
+        assertEquals(pool.stats().busyConnectionCount(), 0);
+        assertEquals(1, pool.stats(). closedConnections());
+    }
+
+    @Test
+    public void closesConnectionAndUpdatesMetricsWhenClosingConnectionFromClosedPool() {
+
+        when(connectionFactory.createConnection(any(Origin.class), any(ConnectionSettings.class)))
+                .thenReturn(Mono.just(connection1));
+
+        ConnectionPoolSettings poolSettings = new ConnectionPoolSettings.Builder()
+                .maxConnectionsPerHost(2)
+                .maxPendingConnectionsPerHost(2)
+                .build();
+
+        SimpleConnectionPool pool = new SimpleConnectionPool(origin, poolSettings, connectionFactory);
+
+        StepVerifier.create(pool.borrowConnection())
+                .expectNext(connection1)
+                .verifyComplete();
+
+        assertEquals(1, pool.stats().busyConnectionCount());
+        //Close connection pool
+        pool.close();
+        pool.closeConnection(connection1);
+
+        verify(connection1).close();
+        assertEquals(pool.stats().terminatedConnections(), 0);
+        assertEquals(pool.stats().closedConnections(), 1);
+        assertEquals(0, pool.stats().busyConnectionCount());
+    }
+
+    @Test
+    public void purgesTerminatedConnectionsForClosedPools() {
+        when(connectionFactory.createConnection(any(Origin.class), any(ConnectionSettings.class)))
+                .thenReturn(Mono.just(connection1));
+
+        // Create a new connection
+        SimpleConnectionPool pool = new SimpleConnectionPool(origin, defaultConnectionPoolSettings(), connectionFactory);
+
+        StepVerifier.create(pool.borrowConnection())
+                .expectNext(connection1)
+                 .verifyComplete();
+        //Close the pool
+        pool.close();
+        //Close and return the connection
+        pool.connectionClosed(connection1);
+        when(connection1.isConnected()).thenReturn(false);
+        pool.returnConnection(connection1);
+
+        assertEquals(pool.stats().connectionAttempts(), 1);
+        assertEquals(pool.stats().pendingConnectionCount(), 0);
+        assertEquals(pool.stats().busyConnectionCount(), 0);
+        assertEquals(pool.stats().availableConnectionCount(), 0);
+
+        assertEquals(pool.stats().terminatedConnections(), 1);
+        assertEquals(pool.stats().closedConnections(), 0);
+    }
+
+
```

Comment:
Good test coverage. However one scenario is omitted: that of a `close()` method closes all active, non-borrowed TCP connections.

Reply:
Added.